collect_app/src/main/java/org/odk/collect/android/widgets/GeoPointWidget.java

```diff
@@ -18,152 +18,174 @@
         widgetValueChanged();
     }
 
-    @Override
-    public String getAnswerToDisplay(String answer) {
-        if (!answer.isEmpty()) {
-            String[] parts = answer.split(" ");
-            return getContext().getString(
-                    R.string.gps_result,
-                    GeoWidgetUtils.convertCoordinatesIntoDegreeFormat(getContext(), Double.parseDouble(parts[0]), "lat"),
-                    GeoWidgetUtils.convertCoordinatesIntoDegreeFormat(getContext(), Double.parseDouble(parts[1]), "lon"),
-                    GeoWidgetUtils.truncateDouble(parts[2]),
-                    GeoWidgetUtils.truncateDouble(parts[3])
-            );
-        } else {
+    private String getAnswerToDisplay(String answer) {
+        try {
+            if (answer != null && !answer.isEmpty()) {
+                String[] parts = answer.split(" ");
+                return getContext().getString(
+                        R.string.gps_result,
+                        GeoWidgetUtils.convertCoordinatesIntoDegreeFormat(getContext(), Double.parseDouble(parts[0]), "lat"),
+                        GeoWidgetUtils.convertCoordinatesIntoDegreeFormat(getContext(), Double.parseDouble(parts[1]), "lon"),
+                        GeoWidgetUtils.truncateDouble(parts[2]),
+                        GeoWidgetUtils.truncateDouble(parts[3])
+                );
+            }
+        } catch (NumberFormatException e) {
             return "";
         }
+        return "";
     }
 
-    @Override
-    public String getDefaultButtonLabel() {
-        return getContext().getString(R.string.get_location);
+    private void updateButtonLabelsAndVisibility(boolean dataAvailable) {
+        if (!readOnly) {
+            binding.simpleButton.setText(
+                    dataAvailable ? R.string.change_location : R.string.get_point);
+        }
     }
 
-    public void startGeoActivity() {
+    private void onButtonClick() {
+        if (MultiClickGuard.allowClick(QuestionWidget.class.getName())) {
+            getPermissionUtils().requestLocationPermissions((Activity) getContext(), new PermissionListener() {
+                @Override
+                public void granted() {
+                    waitingForDataRegistry.waitForData(getFormEntryPrompt().getIndex());
+                    startGeoActivity();
+                }
+
+                @Override
+                public void denied() {
+                }
+            });
+        }
+    }
+
+    private void startGeoActivity() {
         Context context = getContext();
-        Intent intent = new Intent(
-            context, useMap ? GeoPointMapActivity.class : GeoPointActivity.class);
+        Intent intent = new Intent(context, GeoPointActivity.class);
 
         if (stringAnswer != null && !stringAnswer.isEmpty()) {
             intent.putExtra(LOCATION, GeoWidgetUtils.getLocationParamsFromStringAnswer(stringAnswer));
         }
         intent.putExtra(READ_ONLY, readOnly);
```

Comment:
@lognaturel Should I remove `intent.putExtra(READ_ONLY, readOnly)` from here as well.  In case of `readOnly` prompt, the button visibility is set to `View.GONE`. So, the user will not be launching any intent in that case.

Reply:
Good catch, yes!